toolkits/slack/arcade_slack/tools/chat.py

```diff
@@ -72,43 +91,532 @@ def send_dm_to_user(
+    conversation_metadata = await get_conversation_metadata_by_name(
+        context=context, conversation_name=conversation_name
+    )
+    return await get_conversation_history_by_id(
+        context=context,
+        conversation_id=conversation_metadata["id"],
+        oldest_relative=oldest_relative,
+        latest_relative=latest_relative,
+        oldest_datetime=oldest_datetime,
+        latest_datetime=latest_datetime,
+        limit=limit,
+        cursor=cursor,
+    )
+
+
+@tool(
+    requires_auth=Slack(
+        scopes=["channels:read", "groups:read", "im:read", "mpim:read"],
+    )
+)
+async def get_conversation_metadata_by_id(
+    context: ToolContext,
+    conversation_id: Annotated[str, "The ID of the conversation to get metadata for"],
+) -> Annotated[
+    dict,
+    (
+        "The conversation metadata. Metadata does not include: "
+        "- members, users, or people on the conversation / channel; "
+        "- history / messages of the conversation / channel."
+    ),
+]:
+    """Get the metadata of a conversation in Slack."""
+    slackClient = AsyncWebClient(token=context.authorization.token)
+
+    try:
+        response = await slackClient.conversations_info(
+            channel=conversation_id,
+            include_locale=True,
+            include_num_members=True,
         )
 
-    # Step 2: Send the message to the channel
-    response = slackClient.chat_postMessage(channel=channel_id, text=message)
-    response.validate()
+        return extract_conversation_metadata(response["channel"])
+
+    except SlackApiError as e:
+        if e.response["error"] == "channel_not_found":
+            conversations = await list_conversations_metadata(context, limit=-1)
+            available_conversations = ", ".join(
+                f"{conversation['id']} ({conversation['name']})"
+                for conversation in conversations["conversations"]
+            )
+
+            raise RetryableToolError(
+                "Conversation not found",
+                developer_message=f"Conversation with ID '{conversation_id}' not found.",
+                additional_prompt_content=f"Available conversations: {available_conversations}",
+                retry_after_ms=500,
+            )
 
-    if isinstance(response.data, dict):
-        return response.data
-    return {}
+
+@tool(
+    requires_auth=Slack(
+        scopes=["channels:read", "groups:read", "im:read", "mpim:read"],
+    )
+)
+async def get_conversation_metadata_by_name(
+    context: ToolContext,
+    conversation_name: Annotated[str, "The name of the conversation to get metadata for"],
+) -> Annotated[
+    dict,
+    (
+        "The conversation metadata. Metadata does not include: "
+        "- members, users, or people on the conversation / channel; "
+        "- history / messages of the conversation / channel."
+    ),
+]:
+    """Get the metadata of a conversation in Slack."""
+    next_cursor = None
+    conversation_names = []
+
+    while True:
```

Comment:
no. antipattern. relying on breaks scares me. change to a variable and set. ideally with some kind of timeout or secondary fallback to escape the scope.

Reply:
fair. should we use timeouts in all iterations that have the potential to run endlessly?

I can't really enforce a timeout without using threads, though (which is probably undesirable here in terms of perf), but I think a naive timeout implementation would suffice

Reply:
it's done. replaced breaks with a variable check. used async timeout to make sure loops doesn't run past a certain amount of time. since all api calls are asynchronous, we don't actually need a thread to enforce the timeout. thank the async gods 🙏